client/scala/src/main/scala/ai/verta/repository/Commit.scala

```diff
@@ -514,43 +514,122 @@ class Commit(
+                  path_dataset_component_blob_path = Some(datasetComponentPath),
+                  repository_id = Some(VersioningRepositoryIdentification(repo_id = Some(repoId)))
+                )
+              ))
+              .map(_ => ())
+          } finally {
+            inputStream.close()
+          }
+        })
+      }
+      else {
+        Try (new FileInputStream(file)).flatMap(inputStream => { // loan pattern
+          try {
+            Await.result(clientSet.client.requestRaw("PUT", resp.url.get, null, Map("Content-Length" -> file.length.toString), inputStream), Duration.Inf)
+              .map(_ => ())
+          } finally {
+            inputStream.close()
+          }
+        })
+      }
+    })
+  }
+
+  /** Upload a part of the input stream
+   *  @param blobPath path to the blob in the commit
+   *  @param datasetComponentPath path to the component in the blob
+   *  @param inputStream the stream of input to uploaded
+   *  @param partNum the index of the current part
+   *  @return The number of parts uploaded, if succeeds
+   */
+  private def uploadPart(
+    blobPath: String,
+    datasetComponentPath: String,
+    inputStream: InputStream,
+    buffer: Array[Byte],
+    partNum: Int = 1
+  )(implicit ec: ExecutionContext): Try[Int] = {
+    getURLForArtifact(blobPath, datasetComponentPath, "PUT", partNum).map(_.url.get).flatMap(url => {
+      var readLen = inputStream.read(buffer)
+
+      if (readLen < 0)
+        Success(partNum - 1)
+      else
+        retry (
+          Try(new ByteArrayInputStream(buffer)).flatMap(filepart => {
+            try {
+              Await.result(clientSet.client.getRawRequestHeader("PUT", url, null, Map("Content-Length" -> readLen.toString), filepart), Duration.Inf)
+                .flatMap(headers => clientSet.versioningService.commitVersionedBlobArtifactPart(
+                  VersioningCommitVersionedBlobArtifactPart(
+                    artifact_part = Some(ModeldbArtifactPart(
+                      etag = Some(headers("ETag").get),
+                      part_number = Some(BigInt(partNum))
+                    )),
+                    commit_sha = id,
+                    location = Some(pathToLocation(blobPath)),
+                    path_dataset_component_blob_path = Some(datasetComponentPath),
+                    repository_id = Some(VersioningRepositoryIdentification(repo_id = Some(repoId)))
+                  )
+                ))
+                .flatMap(_ => uploadPart(blobPath, datasetComponentPath, inputStream, buffer, partNum + 1))
+            } finally {
+              filepart.close()
+            }
+          }),
+          3, // number of upload attempts
+          f"Uploading part ${partNum} of component ${datasetComponentPath} of blob at ${blobPath} fails."
+        )
+    })
+  }
+
+  /** Helper function to retry failable function.
+   */
+  private def retry[T](f: => Try[T], attempt: Int, errorMessage: String): Try[T] = {
+    if (attempt <= 0)
```

Comment:
```suggestion
    if (attempt < 0)
```
0 just means not retry, right?

Reply:
It's just semantic, but in my current set up 1 would mean not retry. The trace looks sth like this:
- retry(3): If succeeds, done. If fail, call retry(2)
- retry(2): If succeeds, done. If fail, call retry(1)
- retry(1): If succeeds, done. If fail, call retry(0), which fails automaticaly.
So the upload would try for 3 times.

Reply:
Added some documentation to clarify the semantic.

Reply:
Could we change the logic instead of adding documentation? I generally favor strong semantics.

Reply:
Personally I don't have any strong preference. I'll modify this into attemptsLeft. That sounds better right?